vault/logical_system_events.go

```diff
@@ -5,26 +5,209 @@ package vault
 
 import (
 	"context"
+	"fmt"
 	"net/http"
 
+	"github.com/hashicorp/eventlogger"
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault/helper/namespace"
+	"github.com/hashicorp/vault/helper/syncmap"
+	"github.com/hashicorp/vault/plugins/event"
 	"github.com/hashicorp/vault/sdk/framework"
 	"github.com/hashicorp/vault/sdk/logical"
 )
 
+type eventSubscriptions struct {
+	subscriptions *syncmap.SyncMap[string, *eventSubscription]
+}
+
+type eventSubscription struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	id      string
+	plugin  string
+	config  map[string]interface{}
+	backend event.SubscriptionPlugin
+}
+
+func (sub *eventSubscription) ID() string {
+	return sub.id
+}
+
+func newEventSubscriptions() *eventSubscriptions {
+	return &eventSubscriptions{
+		subscriptions: syncmap.NewSyncMap[string, *eventSubscription](),
+	}
+}
+
 // handleEventsSubscribe
-func (b *SystemBackend) handleEventsSubscribe(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-	// TODO
-	return logical.RespondWithStatusCode(nil, req, http.StatusNoContent)
+func (b *SystemBackend) handleEventsSubscribe(requestCtx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	// TODO: check policy
+	eventTypePattern := data.Get("event_type").(string)
+	if eventTypePattern == "" {
+		return logical.ErrorResponse("event_type is required"), logical.ErrInvalidRequest
+	}
+	bexprFilter := data.Get("filter").(string)
+	plugin := data.Get("plugin").(string)
+	if plugin == "" {
+		return logical.ErrorResponse("plugin is required"), logical.ErrInvalidRequest
+	}
+	factory, ok := b.Core.eventBackends[plugin]
+	if !ok {
+		return logical.ErrorResponse("unsupported plugin type"), logical.ErrInvalidRequest
+	}
+	config := data.Get("config").(map[string]interface{})
+	if config == nil || len(config) == 0 {
+		return logical.ErrorResponse("config is required"), logical.ErrInvalidRequest
+	}
+	id, err := uuid.GenerateUUID()
+	if err != nil {
+		b.logger.Error("Error generating UUID", "error", err)
+		return logical.ErrorResponse("error generating UUID"), logical.ErrUnrecoverable
+	}
+	backend, err := factory(requestCtx)
+	if err != nil {
+		b.logger.Error("Error initializing plugin", "error", err)
+		return logical.ErrorResponse("error initializing plugin"), logical.ErrUnrecoverable
+	}
+	if err != nil {
+		b.logger.Warn("Error subscribing to events", "error", err)
+		return logical.ErrorResponse("error subscribing to events"), logical.ErrUnrecoverable
+	}
+	ns, err := namespace.FromContext(requestCtx)
+	if err != nil {
+		b.logger.Error("No namespace found", "error", err)
+		return logical.ErrorResponse("no namespace found"), logical.ErrUnrecoverable
+	}
+	ctx, ctxCancel := context.WithCancel(context.Background())
+	eventCh, eventCancel, err := b.Core.events.Subscribe(ctx, ns, eventTypePattern, bexprFilter)
+	sub := &eventSubscription{
```

Comment:
We should probably return if `err != nil`. It looks like we are not handling the error.